pkg/controller/baremetalhost/host_state_machine.go

```diff
@@ -0,0 +1,294 @@
+	NextState           metal3v1alpha1.ProvisioningState
+	Reconciler          *ReconcileBareMetalHost
+	Provisioner         provisioner.Provisioner
+	RequeueDespiteError bool
+}
+
+func newHostStateMachine(host *metal3v1alpha1.BareMetalHost,
+	reconciler *ReconcileBareMetalHost,
+	provisioner provisioner.Provisioner) *hostStateMachine {
+	currentState := host.Status.Provisioning.State
+	r := hostStateMachine{
+		Host:        host,
+		NextState:   currentState,
+		Reconciler:  reconciler,
+		Provisioner: provisioner,
+	}
+	return &r
+}
+
+type stateHandler func(*reconcileInfo) (reconcile.Result, error)
+
+func (hsm *hostStateMachine) handlers() map[metal3v1alpha1.ProvisioningState]stateHandler {
+	return map[metal3v1alpha1.ProvisioningState]stateHandler{
+		metal3v1alpha1.StateNone:                  hsm.handleNone,
+		metal3v1alpha1.StateRegistering:           hsm.handleRegistering,
+		metal3v1alpha1.StateRegistrationError:     hsm.handleRegistrationError,
+		metal3v1alpha1.StateInspecting:            hsm.handleInspecting,
+		metal3v1alpha1.StateExternallyProvisioned: hsm.handleExternallyProvisioned,
+		metal3v1alpha1.StateMatchProfile:          hsm.handleMatchProfile,
+		metal3v1alpha1.StateReady:                 hsm.handleReady,
+		metal3v1alpha1.StateProvisioning:          hsm.handleProvisioning,
+		metal3v1alpha1.StateProvisioningError:     hsm.handleProvisioningError,
+		metal3v1alpha1.StateProvisioned:           hsm.handleProvisioned,
+		metal3v1alpha1.StatePowerManagementError:  hsm.handlePowerManagementError,
+		metal3v1alpha1.StateDeprovisioning:        hsm.handleDeprovisioning,
+		metal3v1alpha1.StateDeleting:              hsm.handleDeleting,
+	}
+}
+
+func (hsm *hostStateMachine) updateHostStateFrom(initialState metal3v1alpha1.ProvisioningState,
+	result *reconcile.Result,
+	log logr.Logger) {
+	if hsm.NextState != initialState {
+		log.Info("changing provisioning state",
+			"old", initialState,
+			"new", hsm.NextState)
+		hsm.Host.Status.Provisioning.State = hsm.NextState
+		result.Requeue = true
+	}
+}
+
+func (hsm *hostStateMachine) ReconcileState(info *reconcileInfo) (result reconcile.Result, err error) {
+	initialState := hsm.Host.Status.Provisioning.State
+	defer hsm.updateHostStateFrom(initialState, &result, info.log)
+
+	if hsm.shouldInitiateDelete() {
+		info.log.Info("Initiating host deletion")
+		return
+	}
+	// TODO: In future we should always re-register the host if required,
+	// rather than initiate a transistion back to the Registering state.
+	if hsm.shouldInitiateRegister(info) {
+		info.log.Info("Initiating host registration")
+		return
+	}
+
+	if stateHandler, found := hsm.handlers()[initialState]; found {
+		result, err = stateHandler(info)
+	} else {
+		info.log.Info("No handler found for state", "state", initialState)
+		err = fmt.Errorf("No handler found for state \"%s\"", initialState)
+		return
+	}
+
+	return
+}
+
+func (hsm *hostStateMachine) shouldInitiateDelete() bool {
```

Comment:
This function is named in a way that makes me think it answers a question, but it has a side-effect of actually setting up the next state if delete should happen. I don't know if I want to ask to have the function renamed or have the 2 responsibilities divided into separate functions.

Reply:
Will fix in a follow-up. `checkInitiateDelete()` maybe?